### ISSUE TITLE
script coredumps when using threads 

### ISSUE THREAD

**Issue by andreyc21** [ethereum/web3.py#2407]
* Version: 5.28.0 (tried to downgrade to version 5.9, same result )
* Python: 3.10.3 (also reproduces on earlier versions) 
* OS: linux
* `pip freeze` output

```
acme==1.22.0
aiohttp==3.8.1
aiosignal==1.2.0
ansi2html==1.6.0
ansible==2.9.27
appdirs==1.4.4
argcomplete==1.12.3
asgiref==3.4.1
asn1crypto==1.4.0
async-generator==1.10
async-timeout==4.0.2
attrs==21.2.0
Automat==20.2.0
Babel==2.9.1
backcall==0.1.0
base58==2.1.1
bcrypt==3.2.0
Beaker==1.10.0
beautifulsoup4==4.9.3
bitarray==1.2.2
blivet==3.4.3
blivet-gui==2.3.0
Bottleneck==1.3.2
Brlapi==0.8.2
Brotli==1.0.9
cached-property==1.5.2
cairocffi==1.3.0
CairoSVG==2.5.2
certbot==1.22.0
certifi==2020.12.5
cffi==1.15.0
chardet==4.0.0
charset-normalizer==2.0.4
chrome-gnome-shell==0.0.0
click==8.0.1
coincurve==16.0.0
colorama==0.4.4
ConfigArgParse==1.4.1
configobj==5.0.6
confluent-kafka==0.11.6
constantly==15.1.0
cryptography==3.4.7
cssselect2==0.3.0
cupshelpers==1.0
cycler==0.10.0
cytoolz==0.11.2
dasbus==1.6
dateparser==1.1.0
dbus-python==1.2.18
decorator==5.0.9
defusedxml==0.7.1
distro==1.6.0
Django==3.2.12
dnspython==2.1.0
docker==5.0.2
docker-compose==1.29.2
docker-pycreds==0.4.0
docker-squash==1.0.8
dockerpty==0.4.1
docopt==0.6.2
eth-abi==2.1.1
eth-account==0.5.7
eth-hash==0.3.2
eth-keyfile==0.5.1
eth-keys==0.3.4
eth-rlp==0.2.1
eth-typing==2.2.2
eth-utils==1.10.0
ethereum==2.3.2
execnet==1.9.0
fastavro==1.4.9
fedora-third-party==0.8
file-magic==0.4.0
flake8==4.0.1
fonttools==4.26.1
fros==1.1
frozenlist==1.2.0
fs==2.4.11
future==0.18.2
gpg==1.15.1
grafanalib==0.6.2
greenlet==1.1.1
h11==0.12.0
hexbytes==0.2.2
httplib2==0.20.2
humanize==0.5.1
hyperlink==21.0.0
idna==3.2
incremental==21.3.0
iniconfig==1.1.1
initial-setup==0.3.93
ipfshttpclient==0.8.0a2
ipython==7.26.0
ipython_genutils==0.1.0
isc==2.0
jedi==0.18.0
Jinja2==3.0.1
jira==2.0.0
jmespath==0.10.0
josepy==1.9.0
jsonrpcclient==4.0.2
jsonschema==3.2.0
kazoo==2.8.0
kiwisolver==1.4.0
langtable==0.0.57
libcomps==0.1.18
libvirt-python==7.6.0
lru-dict==1.1.7
lxml==4.6.5
Mako==1.1.4
Markdown==3.3.6
MarkupSafe==2.0.0
matplotlib==3.5.1
matplotlib-inline==0.1.2
mccabe==0.6.1
mercurial==5.9.3
multiaddr==0.0.9
multidict==5.2.0
munkres==1.1.2
mypy-extensions==0.4.3
netaddr==0.8.0
nftables==0.1
ntlm-auth==1.5.0
numexpr==2.7.1
numpy==1.21.5
oauthlib==3.0.2
olefile==0.46
outcome==1.1.0
packaging==21.3
pandas==1.3.3
pandas-datareader==0.10.0
paramiko==2.9.1
parsedatetime==2.6
parsimonious==0.8.1
parso==0.8.2
Paste==3.5.0
pbkdf2==1.3
pbr==5.6.0
pexpect==4.8.0
pickleshare==0.7.5
pid==2.2.3
Pillow==8.3.2
pluggy==0.13.1
ply==3.11
productmd==1.33
prometheus-client==0.12.0
prompt-toolkit==3.0.24
protobuf==3.19.1
ptyprocess==0.6.0
pwquality==1.4.4
py==1.11.0
py-ecc==5.2.0
py3nvml==0.2.6
pyasn1==0.4.8
pyasn1-modules==0.2.8
PyAudio==0.2.11
pycairo==1.20.1
pycodestyle==2.8.0
pycparser==2.20
pycrypto==2.6.1
pycryptodome==3.12.0
pycups==2.0.1
pycurl==7.44.1
pyenchant==3.2.2
pyethash==0.1.27
pyflakes==2.4.0
pygame==2.1.2
Pygments==2.9.0
PyGObject==3.42.0
PyJWT==2.1.0
pykickstart==3.34
PyNaCl==1.4.0
pyOpenSSL==21.0.0
pyparsing==2.4.7
pyparted==3.12.0
PyQt5==5.15.6
PyQt5-sip==12.9.0
PyQtWebEngine==5.15.2
pyRFC3339==1.1
pyrsistent==0.18.1
pysha3==1.0.2
PySocks==1.7.1
pystray==0.17.3
pytest==6.2.4
pytest-forked==1.3.0
pytest-xdist==2.3.0
python-augeas==1.1.0
python-binance==1.0.15
python-dateutil==2.8.1
python-dotenv==0.18.0
python-gettext==4.0
python-manatools==0.0.3
python-meh==0.50
python-memcached==1.59
python-xlib==0.31
pytz==2021.3
pytz-deprecation-shim==0.1.0.post0
pyudev==0.22.0
pywinrm==0.4.1
pyxdg==0.27
PyYAML==5.4.1
rancher-gitlab-deploy==1.7
regex==2022.3.2
repoze.lru==0.7
requests==2.27.0
requests-file==1.5.1
requests-ftp==0.3.1
requests-ntlm==1.1.0
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
rlp==1.2.0
rpm==4.17.0
rpmconf==1.1.4
schedule==1.1.0
scipy==1.7.3
scrypt==0.8.19
selenium==4.1.3
selinux==3.3
Send2Trash==1.7.1
sepolicy==3.3
service-identity==21.1.0
setools==4.4.0
setproctitle==1.2.2
shtab==1.5.2
simpleaudio==1.0.4
simplejson==3.17.2
simpleline==1.8.2
six==1.16.0
sniffio==1.2.0
sortedcontainers==2.4.0
sos==4.2
soupsieve==2.3.1
SQLAlchemy==1.4.32
sqlparse==0.4.1
systemd-coredump-python==2
systemd-python==234
tables==3.6.1
Tempita==0.5.2
termcolor==1.1.0
texttable==1.6.2
tinycss2==1.1.1
tldr==3.1.0
toml==0.10.2
toolz==0.11.2
traitlets==5.0.5
trio==0.19.0
trio-websocket==0.9.2
Twisted==21.7.0
typing-extensions==3.10.0.2
tzdata==2021.5
tzlocal==4.1
ujson==5.1.0
urllib3==1.26.7
varint==1.0.2
wcwidth==0.2.5
web3==5.28.0
webencodings==0.5.1
websocket-client==1.2.3
websockets==9.1
wsproto==1.0.0
xcffib==0.11.1
xlrd==2.0.1
xlwt==1.3.0
xmltodict==0.12.0
yarl==1.7.2
zope.component==4.3.0
zope.event==4.2.0
zope.interface==5.4.0
```

With this set of RPCs, the script falls into a coredump on cycles 3-10. 
```
andreyc@andrnote ~/devel $ cat thread5c.py 
import time
from threading import Thread
from web3 import Web3


def run(url, account):
    w3 = Web3(Web3.HTTPProvider(url))
    balance = w3.eth.getBalance(account)
    print("account {}, balance {}".format(account, balance))


addr = "0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d"
while True:
    for url, account in (("https://rpc-mainnet.lachain.io/", addr),
                         ("https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161", addr),
                         ("https://polygon-rpc.com/", addr),
                         ("https://bsc-dataseed.binance.org/", addr),
                         ("https://api.avax.network/ext/bc/C/rpc", addr),
                         ("https://rpc-testnet.lachain.io/", addr),
                         ("https://ropsten.infura.io/v3/9a6feb0c6140449c9997d5f98b2a3665", addr),
                         ("https://matic-mumbai.chainstacklabs.com/", addr),
                         ("https://data-seed-prebsc-1-s2.binance.org:8545/", addr),
                         ("https://api.avax-test.network/ext/bc/C/rpc", addr)):
         Thread(target=run, args=(url, account)).start()
    time.sleep(0.1)
andreyc@andrnote ~/devel $ python thread5c.py 
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 2019591593909653812
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 2019591593909653812
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 1000000000000000000
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 1000000000000000000
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
Exception in thread Thread-54 (run):
account 0xF474Cf03ccEfF28aBc65C9cbaE594F725c80e12d, balance 0
Segmentation fault (core dumped)
```

gdb-py-bt:
```
(gdb) py-bt
Traceback (most recent call first):
  File "/home/andreyc/.local/lib/python3.10/site-packages/eth_utils/address.py", line 118, in <genexpr>
    if int(address_hash[i], 16) > 7
  File "/home/andreyc/.local/lib/python3.10/site-packages/eth_utils/address.py", line 115, in to_checksum_address
    "".join(
  File "/home/andreyc/.local/lib/python3.10/site-packages/eth_utils/address.py", line 134, in is_checksum_address
    is_equal = value == to_checksum_address(value)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/validation.py", line 180, in validate_address
    if not is_checksum_address(value):
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/normalizers.py", line 196, in abi_address_to_hex
    validate_address(data)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/normalizers.py", line 78, in wrapper
    modified = to_wrap(type_str, data)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/abi.py", line 842, in map_to_typed_data
    return ABITypedData(func(*elements))
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/formatters.py", line 90, in recursive_map
    return func(items_mapped)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/decorators.py", line 30, in wrapped
    wrapped_val = to_wrap(*args)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/formatters.py", line 88, in recurse
    return recursive_map(func, item)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/formatters.py", line 76, in map_collection
    return datatype(map(func, collection))
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/formatters.py", line 89, in recursive_map
    items_mapped = map_collection(recurse, data)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/decorators.py", line 30, in wrapped
    wrapped_val = to_wrap(*args)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/abi.py", line 845, in data_tree_map
    return recursive_map(map_to_typed_data, data_tree)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/_utils/abi.py", line 811, in map_abi_data
    return pipe(data, *pipeline)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/formatting.py", line 93, in middleware
    params = formatter(params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/formatting.py", line 94, in middleware
    response = make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/formatting.py", line 94, in middleware
    response = make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/attrdict.py", line 33, in middleware
    response = make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/formatting.py", line 94, in middleware
    response = make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/gas_price_strategy.py", line 90, in middleware
    return make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/middleware/formatting.py", line 94, in middleware
    response = make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/manager.py", line 150, in _make_request
    return request_func(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/manager.py", line 197, in request_blocking
    response = self._make_request(method, params)
  File "/home/andreyc/.local/lib/python3.10/site-packages/web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "/home/andreyc/devel/thread5c.py", line 8, in run
    balance = w3.eth.getBalance(account)
  File "/usr/lib64/python3.10/threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.10/threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.10/threading.py", line 966, in _bootstrap
    self._bootstrap_inner()
```

**Comment by dbfreem** [comment]
I modified this script to run against localhost and let it run for a while. It never did crash. 

Currently one of the reasons that `Eth` is not threadsafe is because of the way sessions are being cached in the `request.py` module. See issue #1847. I made a change to the `AsyncEth` implementation in `request.py` to help it's session caching to be more thread safe but this has not been implemented for `Eth` yet. I was hoping to give the AsyncEth implementation some run time to see if any issues came up before moving it to Eth. 

Here is the change that was made to [AsyncEth's](https://github.com/ethereum/web3.py/issues/2407) request.py implementation. Not, saying this will fix the above reported issue since I couldn't seem to get the crash but it might be part of the issue.

**Comment by andreyc21** [comment]
Thanks! This seems to solve the issue

**Comment by dbfreem** [comment]
@andreyc21 can you take a look at the latest changes in PR #2409 I rearranged the code some and want to make sure it still fixes your issue.

**Comment by andreyc21** [comment]
@dbfreem #2409 also  works stably. Thanks!

**Comment by fselmo** [comment]
Resolved in #2049 for web3.py `v6`. This needs a `v5` backport PR so keeping this open for now.
